buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -0,0 +1,1419 @@
+            "io.netty5.allocator.centralQueueCapacity", NettyRuntime.availableProcessors());
+
+    private final ChunkAllocator chunkAllocator;
+    private final Queue<ChunkByteBuf> centralQueue;
+    private final StampedLock magazineExpandLock;
+    private volatile Magazine[] magazines;
+    private volatile boolean closed;
+
+    AdaptivePoolingAllocator(ChunkAllocator chunkAllocator) {
+        this.chunkAllocator = chunkAllocator;
+        if (javaVersion() < 8) {
+            // The implementation uses StampedLock, which was introduced in Java 8.
+            throw new IllegalStateException("This allocator require Java 8 or newer.");
+        }
+        centralQueue = ObjectUtil.checkNotNull(createSharedChunkQueue(), "centralQueue");
+        magazineExpandLock = new StampedLock();
+        Magazine[] mags = new Magazine[4];
+        for (int i = 0; i < mags.length; i++) {
+            mags[i] = new Magazine(this);
+        }
+        magazines = mags;
+    }
+
+    /**
+     * Create a thread-safe multi-producer, multi-consumer queue to hold chunks that spill over from the
+     * internal Magazines.
+     * <p>
+     * Each Magazine can only hold two chunks at any one time: the chunk it currently allocates from,
+     * and the next-in-line chunk which will be used for allocation once the current one has been used up.
+     * This queue will be used by magazines to share any excess chunks they allocate, so that they don't need to
+     * allocate new chunks when their current and next-in-line chunks have both been used up.
+     * <p>
+     * The simplest implementation of this method is to return a new {@link ConcurrentLinkedQueue}.
+     * However, the {@code CLQ} is unbounded, and this means there's no limit to how many chunks can be cached in this
+     * queue.
+     * <p>
+     * Each chunk in this queue can be up to {@link #MAX_CHUNK_SIZE} in size, so it is recommended to use a bounded
+     * queue to limit the maximum memory usage.
+     * <p>
+     * The default implementation will create a bounded queue with a capacity of {@link #CENTRAL_QUEUE_CAPACITY}.
+     *
+     * @return A new multi-producer, multi-consumer queue.
+     */
+    private static Queue<ChunkByteBuf> createSharedChunkQueue() {
+        return new LinkedBlockingQueue<ChunkByteBuf>(CENTRAL_QUEUE_CAPACITY);
+    }
+
+    public ByteBuf allocate(int size, int maxCapacity) {
+        if (closed) {
+            throw new IllegalStateException("This allocator has been closed.");
+        }
+        if (size <= MAX_CHUNK_SIZE) {
+            AdaptiveByteBuf buf = AdaptiveByteBuf.newInstance();
+            int sizeBucket = AllocationStatistics.sizeBucket(size); // Compute outside of Magazine lock for better ILP.
+            Magazine[] mags;
+            int expansions = 0;
+            do {
+                mags = magazines;
+                int mask = mags.length - 1;
+                int index = (int) (Thread.currentThread().getId() & mask);
```

Comment:
I think `Thread.currentThread().getId()` could be calculated outside of the loop